File: internal/risk/locked_risk.go
package risk

import (
	"context"
	"database/sql"
	"fmt"
	"time"

	"github.com/nonobeam/golang-stock-trading/internal/db/repository"
	"github.com/nonobeam/golang-stock-trading/internal/vn"
)

// LockedRiskCalculator handles locked capital risk calculations for T+2 settlement
type LockedRiskCalculator struct {
	db *sql.DB
}

// NewLockedRiskCalculator creates a new locked risk calculator
func NewLockedRiskCalculator(db *sql.DB) *LockedRiskCalculator {
	return &LockedRiskCalculator{db: db}
}

// LockedRiskResult contains locked risk analysis for a user's portfolio
type LockedRiskResult struct {
	TotalLockedCapital  float64 // Total value of shares in settlement
	TotalLockedRisk     float64 // Worst-case floor-hit risk
	TotalLiquidCapital  float64 // Total value of liquid shares
	TotalLiquidRisk     float64 // Controlled risk (entry - stop)
	AccountValue        float64 // Total account value
	LockedRiskPercent   float64 // Locked risk as % of account
	LockedRiskThreshold float64 // User's configured threshold
	LockedRiskRemaining float64 // Budget remaining for new purchases
	CanAffordNew        bool    // Whether new purchases are allowed
	Positions           []PositionRiskBreakdown
}

// PositionRiskBreakdown shows risk details for a single position
type PositionRiskBreakdown struct {
	Symbol           string
	SettlementStatus string
	LockedCapital    float64
	LockedRisk       float64
	LiquidCapital    float64
	LiquidRisk       float64
	Exchange         string
	DaysUntilLiquid  int
}

// CalculateLockedRisk computes locked risk for shares based on exchange
func CalculateLockedRisk(shares int, price float64, exchange string) float64 {
	capital := float64(shares) * price
	multiplier := vn.GetLockedRiskMultiplierForExchange(exchange)
	return capital * multiplier
}

// CalculateLiquidRisk computes controlled risk for liquid shares
func CalculateLiquidRisk(shares int, entryPrice, stopLoss float64) float64 {
	if entryPrice <= stopLoss {
		return 0
	}
	return float64(shares) * (entryPrice - stopLoss)
}

// GetTotalLockedRisk calculates total locked risk across all positions
func (c *LockedRiskCalculator) GetTotalLockedRisk(ctx context.Context, userID int64) (float64, error) {
	posRepo := repository.NewPositionRepository(c.db)
	return posRepo.GetTotalLockedRisk(ctx, userID)
}

// GetAvailableLockedRiskBudget calculates remaining budget for new locked positions
func (c *LockedRiskCalculator) GetAvailableLockedRiskBudget(ctx context.Context, userID int64, accountValue float64, threshold float64) (float64, error) {
	totalLockedRisk, err := c.GetTotalLockedRisk(ctx, userID)
	if err != nil {
		return 0, err
	}

	maxAllowed := accountValue * threshold
	remaining := maxAllowed - totalLockedRisk

	if remaining < 0 {
		return 0, nil
	}

	return remaining, nil
}

// CanAffordLockedRisk checks if a new purchase fits within locked risk budget
func (c *LockedRiskCalculator) CanAffordLockedRisk(ctx context.Context, userID int64, shares int, price float64, exchange string, accountValue float64, threshold float64) (bool, string, error) {
	// Calculate locked risk for proposed purchase
	newLockedRisk := CalculateLockedRisk(shares, price, exchange)

	// Get current locked risk
	currentLockedRisk, err := c.GetTotalLockedRisk(ctx, userID)
	if err != nil {
		return false, "", err
	}

	// Check against threshold
	maxAllowed := accountValue * threshold
	totalAfterPurchase := currentLockedRisk + newLockedRisk

	if totalAfterPurchase > maxAllowed {
		return false, fmt.Sprintf(
			"Locked risk budget exceeded: current %.0f VND + new %.0f VND = %.0f VND > max %.0f VND (%.0f%% of account)",
			currentLockedRisk,
			newLockedRisk,
			totalAfterPurchase,
			maxAllowed,
			threshold*100,
		), nil
	}

	// Warn if approaching threshold (80%)
	if totalAfterPurchase > maxAllowed*0.8 {
		return true, fmt.Sprintf(
			"WARNING: Locked risk will be %.0f%% of threshold (%.0f / %.0f VND)",
			(totalAfterPurchase/maxAllowed)*100,
			totalAfterPurchase,
			maxAllowed,
		), nil
	}

	return true, "", nil
}

// GetRiskComposition returns breakdown of locked vs liquid risk
func (c *LockedRiskCalculator) GetRiskComposition(ctx context.Context, userID int64) (*LockedRiskResult, error) {
	posRepo := repository.NewPositionRepository(c.db)

	// Get all open positions with settlement data
	positions, err := posRepo.GetOpenPositions(ctx, userID)
	if err != nil {
		return nil, err
	}

	result := &LockedRiskResult{
		Positions: make([]PositionRiskBreakdown, 0),
	}

	currentTime := time.Now()

	for _, pos := range positions {
		breakdown := PositionRiskBreakdown{
			Symbol: pos.Symbol,
		}

		if pos.SettlementStatus != nil {
			breakdown.SettlementStatus = *pos.SettlementStatus
		}

		if pos.Exchange != nil {
			breakdown.Exchange = *pos.Exchange
		} else {
			breakdown.Exchange = vn.GetExchangeFromTicker(pos.Symbol)
		}

		// Calculate locked vs liquid capital
		if pos.IsLocked() {
			if pos.LockedCapital != nil {
				breakdown.LockedCapital = *pos.LockedCapital
			} else {
				breakdown.LockedCapital = float64(pos.Quantity) * pos.EntryPrice
			}
			breakdown.LockedRisk = CalculateLockedRisk(pos.Quantity, pos.EntryPrice, breakdown.Exchange)
			result.TotalLockedCapital += breakdown.LockedCapital
			result.TotalLockedRisk += breakdown.LockedRisk

			// Calculate days until liquid
			if pos.PurchaseDate != nil {
				breakdown.DaysUntilLiquid = vn.GetDaysUntilLiquid(*pos.PurchaseDate, currentTime)
			}
		} else {
			if pos.LiquidCapital != nil {
				breakdown.LiquidCapital = *pos.LiquidCapital
			} else {
				breakdown.LiquidCapital = float64(pos.Quantity) * pos.EntryPrice
			}
			breakdown.LiquidRisk = CalculateLiquidRisk(pos.Quantity, pos.EntryPrice, pos.StopLoss)
			result.TotalLiquidCapital += breakdown.LiquidCapital
			result.TotalLiquidRisk += breakdown.LiquidRisk
		}

		result.Positions = append(result.Positions, breakdown)
	}

	return result, nil
}

// CheckLockedRiskThreshold validates locked risk is within configured limit
func (c *LockedRiskCalculator) CheckLockedRiskThreshold(ctx context.Context, userID int64, accountValue float64, threshold float64) (bool, string, error) {
	currentLockedRisk, err := c.GetTotalLockedRisk(ctx, userID)
	if err != nil {
		return false, "", err
	}

	maxAllowed := accountValue * threshold
	lockedPercent := (currentLockedRisk / accountValue) * 100

	if currentLockedRisk > maxAllowed {
		return false, fmt.Sprintf(
			"Locked risk threshold exceeded: %.0f VND (%.2f%%) > max %.0f VND (%.0f%%)",
			currentLockedRisk,
			lockedPercent,
			maxAllowed,
			threshold*100,
		), nil
	}

	// Warn if above 80% of threshold
	if currentLockedRisk > maxAllowed*0.8 {
		return true, fmt.Sprintf(
			"WARNING: Locked risk at %.0f%% of threshold (%.0f / %.0f VND)",
			(currentLockedRisk/maxAllowed)*100,
			currentLockedRisk,
			maxAllowed,
		), nil
	}

	return true, "", nil
}

// CalculateMaxSharesForLockedRiskBudget determines max shares purchasable within budget
func (c *LockedRiskCalculator) CalculateMaxSharesForLockedRiskBudget(ctx context.Context, userID int64, price float64, exchange string, accountValue float64, threshold float64) (int, error) {
	availableBudget, err := c.GetAvailableLockedRiskBudget(ctx, userID, accountValue, threshold)
	if err != nil {
		return 0, err
	}

	if availableBudget <= 0 {
		return 0, nil
	}

	// Calculate max capital we can lock
	multiplier := vn.GetLockedRiskMultiplierForExchange(exchange)
	maxCapital := availableBudget / multiplier

	// Convert to shares
	maxShares := int(maxCapital / price)

	// Round down to lot size
	return RoundToLotSize(float64(maxShares), GetLotSize()), nil
}

File: ml-service/position_sizing/locked_risk.py
"""
Locked Risk Calculator for T+2 Settlement

Calculates worst-case floor-hit risk for shares in settlement period.
Vietnamese market: shares purchased cannot be sold for 3 trading days (T+0, T+1, T+2 -> sellable on T+3).
"""

import psycopg3
from typing import Dict, Tuple, Optional
from datetime import datetime


class LockedRiskCalculator:
    """Calculates and validates locked capital risk during T+2 settlement period."""

    # Exchange-specific risk multipliers (floor limit + margin for slippage/fees)
    EXCHANGE_RISK_MULTIPLIERS = {
        'HOSE': 0.20,   # 7% floor + margin = 20% worst case
        'HNX': 0.30,    # 10% floor + margin = 30% worst case
        'UPCOM': 0.40,  # 15% floor + margin = 40% worst case
    }

    def __init__(self, db_conn: psycopg3.Connection):
        """
        Initialize locked risk calculator.

        Args:
            db_conn: PostgreSQL database connection
        """
        self.db = db_conn

    def get_exchange_risk_multiplier(self, exchange: str) -> float:
        """
        Get risk multiplier for an exchange.

        Args:
            exchange: Exchange code (HOSE, HNX, UPCOM)

        Returns:
            Risk multiplier (0.20 for HOSE, 0.30 for HNX, 0.40 for UPCOM)
        """
        return self.EXCHANGE_RISK_MULTIPLIERS.get(exchange, 0.20)

    def calculate_locked_risk(self, shares: int, price: float, exchange: str) -> float:
        """
        Calculate worst-case locked risk for shares in settlement.

        Args:
            shares: Number of shares
            price: Entry price per share
            exchange: Exchange code

        Returns:
            Locked risk amount in VND
        """
        capital = shares * price
        multiplier = self.get_exchange_risk_multiplier(exchange)
        return capital * multiplier

    def get_total_locked_risk(self, user_id: int) -> float:
        """
        Calculate total locked risk across all user's locked positions.

        Args:
            user_id: User ID

        Returns:
            Total locked risk in VND
        """
        with self.db.cursor() as cur:
            cur.execute("""
                SELECT COALESCE(
                    SUM(
                        CASE
                            WHEN exchange = 'HOSE' THEN locked_capital * 0.20
                            WHEN exchange = 'HNX' THEN locked_capital * 0.30
                            WHEN exchange = 'UPCOM' THEN locked_capital * 0.40
                            ELSE locked_capital * 0.20
                        END
                    ),
                    0
                )
                FROM positions
                WHERE user_id = %s
                    AND settlement_status IN ('LOCKED_T0', 'LOCKED_T1', 'LOCKED_T2')
                    AND is_closed = FALSE
            """, (user_id,))

            result = cur.fetchone()
            return float(result[0]) if result else 0.0

    def get_locked_risk_budget(self, user_id: int, account_value: float, threshold: float) -> Dict[str, float]:
        """
        Get locked risk budget status.

        Args:
            user_id: User ID
            account_value: Total account value
            threshold: Locked risk threshold (e.g., 0.10 for 10%)

        Returns:
            Dict with total_locked_risk, max_allowed, available, used_percent
        """
        total_locked_risk = self.get_total_locked_risk(user_id)
        max_allowed = account_value * threshold
        available = max(0, max_allowed - total_locked_risk)
        used_percent = (total_locked_risk / max_allowed * 100) if max_allowed > 0 else 0

        return {
            'total_locked_risk': total_locked_risk,
            'max_allowed': max_allowed,
            'available': available,
            'used_percent': used_percent,
            'threshold_percent': threshold * 100
        }

    def check_locked_risk_budget(
        self,
        user_id: int,
        ticker: str,
        shares: int,
        price: float,
        account_value: float,
        threshold: float = 0.10
    ) -> Tuple[bool, Optional[str]]:
        """
        Check if a purchase fits within locked risk budget.

        Args:
            user_id: User ID
            ticker: Stock ticker
            shares: Number of shares to purchase
            price: Entry price
            account_value: Total account value
            threshold: Locked risk threshold (default 10%)

        Returns:
            Tuple of (approved, rejection_message)
            - approved: True if purchase is allowed
            - rejection_message: Reason if rejected, None if approved
        """
        # Get exchange from ticker (simplified - should use lookup table in production)
        exchange = self._get_exchange_from_ticker(ticker)

        # Calculate locked risk for new purchase
        new_locked_risk = self.calculate_locked_risk(shares, price, exchange)

        # Get current locked risk
        current_locked_risk = self.get_total_locked_risk(user_id)

        # Check against threshold
        max_allowed = account_value * threshold
        total_after_purchase = current_locked_risk + new_locked_risk

        if total_after_purchase > max_allowed:
            message = (
                f"Locked risk budget exceeded: "
                f"current {current_locked_risk:,.0f} VND + "
                f"new {new_locked_risk:,.0f} VND = "
                f"{total_after_purchase:,.0f} VND > "
                f"max {max_allowed:,.0f} VND ({threshold*100:.0f}% of account)"
            )
            return False, message

        # Warn if approaching threshold (80%)
        if total_after_purchase > max_allowed * 0.8:
            warning = (
                f"WARNING: Locked risk will be {total_after_purchase/max_allowed*100:.0f}% "
                f"of threshold ({total_after_purchase:,.0f} / {max_allowed:,.0f} VND)"
            )
            return True, warning

        return True, None

    def _get_exchange_from_ticker(self, ticker: str) -> str:
        """
        Infer exchange from ticker symbol.

        This is a simplified heuristic. In production, use actual ticker-to-exchange mapping.

        Args:
            ticker: Stock ticker

        Returns:
            Exchange code (HOSE, HNX, or UPCOM)
        """
        # TODO: Replace with actual exchange lookup from database
        # For now, default to HOSE (most common)
        return 'HOSE'

    def calculate_max_shares_for_budget(
        self,
        user_id: int,
        ticker: str,
        price: float,
        account_value: float,
        threshold: float = 0.10
    ) -> int:
        """
        Calculate maximum shares purchasable within locked risk budget.

        Args:
            user_id: User ID
            ticker: Stock ticker
            price: Entry price per share
            account_value: Total account value
            threshold: Locked risk threshold

        Returns:
            Maximum number of shares (rounded down to lot size of 100)
        """
        # Get available budget
        budget_info = self.get_locked_risk_budget(user_id, account_value, threshold)
        available = budget_info['available']

        if available <= 0:
            return 0

        # Get exchange and multiplier
        exchange = self._get_exchange_from_ticker(ticker)
        multiplier = self.get_exchange_risk_multiplier(exchange)

        # Calculate max capital we can lock
        max_capital = available / multiplier

        # Convert to shares
        max_shares = int(max_capital / price)

        # Round down to lot size (100 shares in Vietnamese market)
        lot_size = 100
        return (max_shares // lot_size) * lot_size


def get_user_locked_risk_threshold(db_conn: psycopg3.Connection, user_id: int) -> float:
    """
    Get user's configured locked risk threshold from database.

    Args:
        db_conn: Database connection
        user_id: User ID

    Returns:
        Locked risk threshold (default 0.10 if not set)
    """
    with db_conn.cursor() as cur:
        cur.execute("""
            SELECT COALESCE(locked_risk_threshold, 0.10)
            FROM user_config
            WHERE user_id = %s
        """, (user_id,))

        result = cur.fetchone()
        return float(result[0]) if result else 0.10
